WordPress/gutenberg
# [RNMobile] Bottom-sheet: Add custom header

## Description


Add a `header` prop for rendering a custom header in the bottom-sheet component. This PR also introduces fixes to address the [issue](https://github.com/WordPress/gutenberg/issues/31151) related to the bottom-sheet component and the keyboard show/hide animation.

Additionally, it also modifies how the inserter search form is rendered in the inserter menu, now it's rendered via the new `header` prop.

## How has this been tested?



1. Open the inserter menu
2. Observe that the inserter search input is rendered properly and that the blocks list can be navigated
3. Tap on the search input
4. Observe that the animation behaves properly
5. Tap on the back arrow to close the keyboard
6. Observe that the animation behaves properly

### Check header changes upon keyboard show/hide
As an additional test, I applied the changes from the patch diff to verify if making changes in the header when the keyboard is shown/hide could produce visual glitches.

<details><summary>Patch diff</summary>

```diff
diff --git a/packages/block-editor/src/components/inserter/menu.native.js b/packages/block-editor/src/components/inserter/menu.native.js
index a81cfbd184db8005d4814b8576a7d0dc4e2a3f7a..46d1778f7a5c0992be66100d9b8f353404869e9b 100644
--- a/packages/block-editor/src/components/inserter/menu.native.js
+++ b/packages/block-editor/src/components/inserter/menu.native.js
@@ -41,6 +41,7 @@ function InserterMenu( {
 	const [ filterValue, setFilterValue ] = useState( '' );
 	// eslint-disable-next-line no-undef
 	const [ showSearchForm, setShowSearchForm ] = useState( __DEV__ );
+	const [ showExtraHeader, setShowExtraHeader ] = useState( true );
 
 	const {
 		showInsertionPoint,
@@ -180,15 +181,30 @@ function InserterMenu( {
 		<BottomSheet
 			isVisible={ true }
 			onClose={ onClose }
+			onKeyboardShow={ () => setShowExtraHeader( false ) }
+			onKeyboardHide={ () => setShowExtraHeader( true ) }
 			header={
-				showSearchForm && (
-					<InserterSearchForm
-						onChange={ ( value ) => {
-							setFilterValue( value );
-						} }
-						value={ filterValue }
-					/>
-				)
+				<>
+					{ showSearchForm && (
+						<InserterSearchForm
+							onChange={ ( value ) => {
+								setFilterValue( value );
+							} }
+							value={ filterValue }
+						/>
+					) }
+					{ showExtraHeader && (
+						<View
+							style={ {
+								marginHorizontal: 8,
+								marginBottom: 20,
+								height: 40,
+								borderRadius: 8,
+								backgroundColor: 'black',
+							} }
+						/>
+					) }
+				</>
 			}
 			hasNavigation
 			setMinHeightToMaxHeight={ showSearchForm }
```

</details>

1. Open the inserter menu
2. Observe that the inserter search input is rendered properly and that the blocks list can be navigated
3. Tap on the search input
4. Observe that the animation behaves properly
5. Tap on the back arrow to close the keyboard
6. Observe that the animation behaves properly

## Screenshots 

iOS|iOS|Android|Android
:---:|:---:|:---:|:---:
Static header|Header modified|Static header|Header modified
<video src=https://user-images.githubusercontent.com/14905380/116706666-22f87f00-a9ce-11eb-89b4-3950f0c7be41.mp4>|<video src=https://user-images.githubusercontent.com/14905380/116706681-27249c80-a9ce-11eb-9728-b4c87fb8bb9c.mp4>|<video src=https://user-images.githubusercontent.com/14905380/116706690-2ab82380-a9ce-11eb-9e9d-e4fc6e443b19.mp4>|<video src=https://user-images.githubusercontent.com/14905380/116706697-2c81e700-a9ce-11eb-826b-81ed9476f727.mp4>

**As you can see on the videos, on Android the animation is different because we can't listen to the `will` events of the keyboard.**

## Types of changes




Bug fix

## Checklist:
